src/components/FlatTreeRenderer.js

```diff
@@ -0,0 +1,436 @@
+    heightObjectRef.current = { ...heightObjectRef.current, ...updatedHeightObject }
+    const updatedOffsetObj = calculateYoffset(heightObjectRef.current, flatArrayRef.current)
+    setYO(updatedOffsetObj)
+  }, [])
+
+  return { yOffsetObject, heightObj: heightObjectRef, setHeightObject }
+}
+
+/**
+ * Component that handles flat tree group animations.
+ */
+const TreeAnimation = ({
+  flatArray,
+  flatArrayKey,
+  visibleStartDepth,
+  oldFlatArray,
+  oldFlatArrayKey
+}) => {
+  const { yOffsetObject, heightObj, setHeightObject } = useYoffsetHandler(flatArray)
+  const yOffsetObjectOldRef = useRef(yOffsetObject)
+  const heightObjectPendingRef = useRef([])
+
+  useEffect(() => {
+    yOffsetObjectOldRef.current = yOffsetObject
+  }, [yOffsetObject])
+
+  const yOffsetObjectOld = yOffsetObjectOldRef.current
+
+  const transitions = useTransition(
+    flatArray,
+    node => node.key,
+    {
+      config: SPRING_CONFIG_GROUP,
+      unique: true,
+      from: {
+        opacity: 0
+      },
+      enter: i => async next => {
+        const item = flatArrayKey[i.key] || i
+        wait(20)
+        // find the item just above it in the flat array which was available in old flat array
+        const lastItem = flatArray[Math.max(0, item.index - 1)]
+        const yOffsetObject = calculateYoffset({ ...heightObj.current }, flatArray)
+        const yOffsetStart = yOffsetObject[lastItem.key]
+        const yOffsetTo = yOffsetObject[item.key]
+
+        const isSecondColumn = item.viewInfo.table.column === 2
+
+        const xOffsetTo = calculateXOffset(item, visibleStartDepth)
+        const xOffsetStart = isSecondColumn ? xOffsetTo - 10 : calculateXOffset(lastItem, visibleStartDepth)
+
+        await next({
+          opacity: 0,
+          x: `${xOffsetStart > xOffsetTo ? xOffsetTo : xOffsetStart}rem`,
+          ...!isNaN(yOffsetStart) ? { y: `${yOffsetStart}px` } : {}
+        })
+
+        await next({
+          opacity: 1,
+          rotation: item.expanded ? 90 : 0,
+          x: `${xOffsetTo}rem`,
+          selectionOpacity: item.isCursor ? TEXT_SELECTION_OPCAITY : 0,
+          ...!isNaN(yOffsetTo) ? { y: `${yOffsetTo}px` } : {}
+        })
+      },
+      leave: item => async next => {
+        // we need new nodes height for animating out y offset animations
+        // waiting for heightObject update
+        await wait(20)
+
+        const isSecondColumn = item.viewInfo.table.column === 2
+
+        // find the item which is available in new flat array and was above the leaving node in old array
+        const lastItem = oldFlatArray.slice(0, item.index).reduce(
+          (acc, _item) => {
+            const isSibling = _item.path.length === item.path.length && isDescendant(contextOf(pathToContext(_item.path)), pathToContext(item.path))
+            return flatArrayKey[_item.key] && (isDescendant(pathToContext(_item.path), pathToContext(item.path)) || isSibling) ? _item : acc
+          },
+          null)
```

Comment:
Using the ordering of flatArray as an implicit API to determine parents and siblings is a bit of a red flag for me. Noting for later review.